Refactor checks whether connections should be established

jougs opened the pull request:
This PR removes the checks for whether a connection should be established between two nodes from the individual connection routines and adds them again as helper function `connection_required()` to the `ConnectionManager`. Additionally, the checks are now cleaned up and better documented.

**Please note** that ten topology tests (1 plain, 9 with MPI) are still failing with this PR in place. As it stands (at least to my current knowledge), they have nothing to do with the connection checks as this refactoring did not change anything for them and I really have no clue what else the checks could  be doing wrong at this point.

I understand that the investigation of this is extremely tedious and time consuming. If all else fails, we could have a code staring VC next week.

jougs committed:
Adapt tarball name to GitHub convention

heplesser committed:
Merge pull request #852 from jougs/update-install-doc

Adapt tarball name in documentation to GitHub convention

jougs committed:
Merge branch 'master' of github.com:compneuronmbu/nest-simulator into compneuronmbu/new-conn-checker

jougs committed:
Refactor checks whether connections should be established

This commit removes the checks for whether a connection should be
established between two nodes from the individual connection routines
and adds them again as function ConnectionManager::connection_required().
Additionally, the checks are now cleaned up and better documented.

jougs committed:
Merge branch 'fix-481-SubnetFreeTopology' into compneuronmbu/new-conn-checker

hakonsbm commented:
It looks like the commit causing the trouble with the 9 MPI tests is a381472f2215bbf2acea418e9911f8af583cc704, and specifically this: 
```c++
if ( not kernel().vp_manager.is_local_vp( vp ) )
{
  return 0;
}
```
If the vp isn't local, shouldn't the function return a proxy node instead, like in [the next return statement](https://github.com/compneuronmbu/nest-simulator/commit/a381472f2215bbf2acea418e9911f8af583cc704#diff-371fef123b62af2b34f8800d45f346ebR420)?

jougs committed:
Return a proxy also for remote nodes

This makes the single-argument NodeManager::get_node_or_proxy() return
a proxy also for remote nodes. As the thread is not correctly set in
this case, I also added an assertion that fails in case the thread of
a proxy is read out.

jougs committed:
Fix missing connections for Topology's target_driven_connect

jougs commented:
@hakonsbm: Many thanks for the hint. In 443a5c8 I've changed the code to return a proxy also in the case of an MPI-remote node, but decided to protect the call to `proxynode::get_thread()` by an assertion, as the thread number will be set wrongly in that case. I'll again look into the removal of all dangerous calls to `get_node_or_proxy()` soon.

I've also found and fixed the problem with the one remaining failing test in `target_driven_connect`.

This means that the branch now builds and tests without failures. Yay!
